subxt/src/backend/rpc/reconnecting_rpc_client/mod.rs

```diff
@@ -0,0 +1,630 @@
+// Copyright 2019-2024 Parity Technologies (UK) Ltd.
+// This file is dual-licensed as Apache-2.0 or GPL-3.0.
+// see LICENSE for license details.
+
+//! # reconnecting-jsonrpsee-ws-client
+//!
+
+mod platform;
+#[cfg(test)]
+mod tests;
+mod utils;
+
+use std::{
+    pin::Pin,
+    sync::Arc,
+    task::{self, Poll},
+    time::Duration,
+};
+
+use super::{RawRpcFuture, RawRpcSubscription, RpcClientT};
+use crate::error::RpcError as SubxtRpcError;
+
+use finito::Retry;
+use futures::{Future, FutureExt, Stream, StreamExt, TryStreamExt};
+use jsonrpsee::core::{
+    client::{
+        Client as WsClient, ClientT, Subscription as RpcSubscription, SubscriptionClientT,
+        SubscriptionKind,
+    },
+    traits::ToRpcParams,
+};
+use serde_json::value::RawValue;
+use tokio::sync::{
+    mpsc::{self, UnboundedReceiver, UnboundedSender},
+    oneshot, Notify,
+};
+use utils::display_close_reason;
+use utils::{reconnect_channel, ReconnectRx, ReconnectTx};
+
+// re-exports
+pub use finito::{ExponentialBackoff, FibonacciBackoff, FixedInterval};
+pub use jsonrpsee::core::client::IdKind;
+pub use jsonrpsee::{core::client::error::Error as RpcError, rpc_params, types::SubscriptionId};
+
+#[cfg(feature = "native")]
+pub use jsonrpsee::ws_client::{HeaderMap, PingConfig};
+
+const LOG_TARGET: &str = "subxt-reconnecting-rpc-client";
+
+/// Method result.
+pub type MethodResult = Result<Box<RawValue>, Error>;
+/// Subscription result.
+pub type SubscriptionResult = Result<Box<RawValue>, DisconnectedWillReconnect>;
+
+/// The connection was closed, reconnect initiated and the subscription was dropped.
+#[derive(Debug, thiserror::Error)]
+#[error("The connection was closed because of `{0:?}` and reconnect initiated")]
+pub struct DisconnectedWillReconnect(String);
+
+/// Serialized JSON-RPC params.
+#[derive(Debug, Clone)]
+pub struct RpcParams(Option<Box<RawValue>>);
```

Comment:
No need for this to be public, since it's only used internally?

Reply:
It's needed for jsonrpsee under the hood, we need something that implements ToRpcParams but yeah no need to make it public.